platform/wayland/cog-platform-wl.c

```diff
@@ -2236,124 +2386,76 @@ create_popup (WebKitOptionMenu *option_menu)
-                                  &xdg_surface_listener,
-                                  NULL);
-        popup_data.xdg_popup = xdg_surface_get_popup (popup_data.xdg_surface,
-                                                      win_data.xdg_surface,
-                                                      popup_data.xdg_positioner);
-        g_assert (popup_data.xdg_popup);
-
-        xdg_popup_add_listener (popup_data.xdg_popup,
-                                &xdg_popup_listener,
-                                NULL);
-        xdg_popup_grab (popup_data.xdg_popup, wl_data.seat, wl_data.event_serial);
-        wl_surface_commit (popup_data.wl_surface);
-    } else if (wl_data.shell != NULL) {
-        popup_data.shell_surface = wl_shell_get_shell_surface (wl_data.shell,
-                                                               popup_data.wl_surface);
-        g_assert(popup_data.shell_surface);
-
-        wl_shell_surface_add_listener (popup_data.shell_surface,
-                                       &shell_popup_surface_listener,
-                                       NULL);
-        wl_shell_surface_set_popup (popup_data.shell_surface,
-                                    wl_data.seat, wl_data.event_serial,
-                                    win_data.wl_surface,
-                                    0, (win_data.height - popup_data.height), 0);
+    popup_data.height = cog_popup_menu_get_height_for_option_menu(option_menu);
 
-        display_popup();
-    }
+    popup_data.popup_menu = cog_popup_menu_create(option_menu, wl_data.shm, popup_data.width, popup_data.height,
+                                                  wl_data.current_output.scale);
+
+    popup_data.shell_context->wl_shell_data.wl_surface = wl_compositor_create_surface(wl_data.compositor);
+    g_assert(popup_data.shell_context->wl_shell_data.wl_surface);
+    wl_surface_set_buffer_scale(popup_data.shell_context->wl_shell_data.wl_surface, wl_data.current_output.scale);
+    win_data.shell_context->functions.create_popup();
 }
 
 static void
-destroy_popup (void)
+destroy_popup(void)
 {
     if (popup_data.option_menu == NULL)
         return;
 
-    webkit_option_menu_close (popup_data.option_menu);
-    g_clear_pointer (&popup_data.popup_menu, cog_popup_menu_destroy);
-    g_clear_object (&popup_data.option_menu);
+    webkit_option_menu_close(popup_data.option_menu);
+    g_clear_pointer(&popup_data.popup_menu, cog_popup_menu_destroy);
+    g_clear_object(&popup_data.option_menu);
 
-    g_clear_pointer (&popup_data.xdg_popup, xdg_popup_destroy);
-    g_clear_pointer (&popup_data.xdg_surface, xdg_surface_destroy);
-    g_clear_pointer (&popup_data.xdg_positioner, xdg_positioner_destroy);
-    g_clear_pointer (&popup_data.shell_surface, wl_shell_surface_destroy);
-    g_clear_pointer (&popup_data.wl_surface, wl_surface_destroy);
+    g_clear_pointer(&popup_data.shell_context->xdg_shell_data.xdg_popup, xdg_popup_destroy);
```

Comment:
Same with this block, you can't access multiple members of `shell_context` like this, only one is going to be valid.